### PR TITLE
introduce pull policy to docker images

### PR DESCRIPTION
**Pull request by TomasTomecek**
TODO:
* [x] docs
* [x] examples
* [x] a bit more testing

I'm open to all the comments.

### CODE REVIEW

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: success, return code = __0__ 
* Result url: __[https://da.gd/pixRr](https://da.gd/pixRr)__
* Comment: __See details for logs__

File: conu/backend/docker/image.py [hunk 1]
```diff
@@ -82,23 +102,25 @@ def get_id(self):
...
 
-        if present and not always:
-            return
+    def pull(self):
+        """
+        Pull this image from registry.
 
+        :return: None
+        """
+        # pull doesn't throw an exc if it fails, should we?
```
Comment (dhodovsk) [inline conu/backend/docker/image.py]: +1, e.g. for early catching of potential issues

File: tests/integration/test_docker.py [hunk 2]
```diff
@@ -220,3 +220,30 @@ def test_nonblocking_execute():
...
+    with DockerBackend() as backend:
+        image = backend.ImageClass("docker.io/library/busybox", tag="1.25.1")
+        try:
+            assert image.is_present()
+        finally:
+            image.rmi(force=True)
+
+
+def test_pull_never():
+    with DockerBackend() as backend:
```
Comment (dhodovsk) [inline tests/integration/test_docker.py]: please add `assert not image.is_present()` also before the call so when this fails we know that it is cause of functionality of image __init__, not presence of busybox on localhost.

File: conu/backend/docker/image.py [hunk 3]
```diff
@@ -44,19 +46,37 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         return super(DockerImageFS, self).__exit__(exc_type, exc_val, exc_tb)
 
 
+class DockerImagePullPolicy(enum.Enum):
+    NEVER = 0
+    IF_NOT_PRESENT = 1
+    ALWAYS = 2
```
Comment (dhodovsk) [inline conu/backend/docker/image.py]: One can find in ALWAYS policy promise that the actual pulling would happen every time (not just contacting docker API). We might want to specify in documentation why pulling image, that is already present, wouldn't be beneficial so there is no confusion.

File: conu/backend/docker/image.py [hunk 4]
```diff
@@ -44,19 +46,37 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+
+
 class DockerImage(Image):
     """
     Utility functions for docker images.
     """
-    def __init__(self, repository, tag="latest"):
+    def __init__(self, repository, tag="latest", pull_policy=DockerImagePullPolicy.IF_NOT_PRESENT):
```
Comment (dhodovsk) [inline conu/backend/docker/image.py]: Not important, but - Is there a possibility to early catch this bad example of usage?
```
>>> i = DockerBackend.ImageClass('fedora', DockerImagePullPolicy.ALWAYS)
08:33:00.186 image.py          DEBUG  pull policy set to 'if_not_present' and image is not present, pulling the image
08:33:00.215 image.py          DEBUG  b'{"status":"Trying to pull repository registry.fedoraproject.org/fedora ... "}\r\n'
08:33:02.399 image.py          DEBUG  b'{"status":"Trying to pull repository registry.access.redhat.com/fedora ... "}\r\n'
08:33:04.454 image.py          DEBUG  b'{"status":"Trying to pull repository docker.io/library/fedora ... "}\r\n'
08:33:06.330 image.py          DEBUG  b'{"errorDetail":{"message":"manifest for docker.io/fedora:DockerImagePullPolicy.ALWAYS not found"},"error":"manifest for docker.io/fedora:DockerImagePullPolicy.ALWAYS not found"}\r\n'
```

**Comment by psklenar** [comment]
## jenkins/superCI comment
* Status: success, return code = __0__ 
* Result url: __[https://da.gd/Lhzo](https://da.gd/Lhzo)__
* Comment: __See details for logs__

**Comment by dhodovsk** [comment]
I was just thinking, maybe we could somehow associate image presence check and pull with docker api client usage. That would cover all methods, aside from _run_container, mount and S2IDockerImage extend and usage. However, it might get difficult with other methods incoming.

**Comment by TomasTomecek** [comment]
@dhodovsk sounds good!

**Pull request merged**